### PR TITLE
CASSANDRA-18941: Support max SSTable size in sorted CQLSSTableWriter

### PR DESCRIPTION
**Pull request by yifan-c**
Thanks for sending a pull request! Here are some tips if you're new here:
 
 * Ensure you have added or run the [appropriate tests](https://cassandra.apache.org/_/development/testing.html) for your PR.
 * Be sure to keep the PR description updated to reflect all changes.
 * Write your PR title to summarize what this PR proposes.
 * If possible, provide a concise example to reproduce the issue for a faster review.
 * Read our [contributor guidelines](https://cassandra.apache.org/_/development/index.html)
 * If you're making a documentation change, see our [guide to documentation contribution](https://cassandra.apache.org/_/development/documentation.html)
 
Commit messages should follow the following format:

```
<One sentence description, usually Jira title or CHANGES.txt summary>

<Optional lengthier description (context on patch)>

patch by <Authors>; reviewed by <Reviewers> for CASSANDRA-#####

Co-authored-by: Name1 <email1>
Co-authored-by: Name2 <email2>

```

The [Cassandra Jira](https://issues.apache.org/jira/projects/CASSANDRA/issues/)

### CODE REVIEW

File: src/java/org/apache/cassandra/io/sstable/CQLSSTableWriter.java [hunk 1]
```diff
@@ -482,36 +486,56 @@ public Builder using(String modificationStatement)
...
+         *
+         * @param size the maximum sizein mebibytes of each individual SSTable allowed
+         * @return this builder
+         */
+        public Builder withMaxSSTableSizeInMiB(int size)
+        {
+            if (size <= 0)
+            {
+                logger.warn("A non-positive value for maximum SSTable size is specified, " +
+                        "which disables the size limiting effectively. Please supply a positive value in order " +
```
Comment (Maxwell-Guo) [inline src/java/org/apache/cassandra/io/sstable/CQLSSTableWriter.java]: code format

File: test/unit/org/apache/cassandra/io/sstable/CQLSSTableWriterTest.java [hunk 2]
```diff
@@ -1116,6 +1117,78 @@ public void testFrozenSetTypeCustomOrdered() throws Exception
...
     }
 
+    @Test
+    public void testWriteWithSorted() throws Exception
+    {
+        String schema = "CREATE TABLE " + qualifiedTable + " ("
+                + "  k int PRIMARY KEY,"
+                + "  v blob )";
+        CQLSSTableWriter writer = CQLSSTableWriter.builder()
+                .inDirectory(dataDir)
```
Comment (Maxwell-Guo) [inline test/unit/org/apache/cassandra/io/sstable/CQLSSTableWriterTest.java]: code format

File: test/unit/org/apache/cassandra/io/sstable/CQLSSTableWriterTest.java [hunk 3]
```diff
@@ -1116,6 +1117,78 @@ public void testFrozenSetTypeCustomOrdered() throws Exception
...
+        }
+    }
+
+    @Test
+    public void testWriteWithSortedAndMaxSize() throws Exception {
+        String schema = "CREATE TABLE " + qualifiedTable + " ("
+                + "  k int PRIMARY KEY,"
+                + "  v blob )";
+        CQLSSTableWriter writer = CQLSSTableWriter.builder()
+                .inDirectory(dataDir)
```
Comment (Maxwell-Guo) [inline test/unit/org/apache/cassandra/io/sstable/CQLSSTableWriterTest.java]: code format

File: test/unit/org/apache/cassandra/io/sstable/CQLSSTableWriterTest.java [hunk 4]
```diff
@@ -1116,6 +1117,78 @@ public void testFrozenSetTypeCustomOrdered() throws Exception
...
+        // 30_000 rows should take 30_000 * (4 + 37) = 1.17 MiB > 1 MiB
+        for (int i = 0; i < rowCount; i++) {
+            writer.addRow(i, UUID.randomUUID().toString());
+        }
+        writer.close();
+
+        File[] dataFiles = dataDir.list(f -> f.name().endsWith(Component.DATA.type.repr));
+        assertNotNull(dataFiles);
+        assertEquals("The sorted writer should produce 2 sstables when max sstable size is configured",
+                2, dataFiles.length);
```
Comment (Maxwell-Guo) [inline test/unit/org/apache/cassandra/io/sstable/CQLSSTableWriterTest.java]: code format

File: test/unit/org/apache/cassandra/io/sstable/CQLSSTableWriterTest.java [hunk 5]
```diff
@@ -1116,6 +1117,78 @@ public void testFrozenSetTypeCustomOrdered() throws Exception
...
+        long closeTo1MiBFileSize = Math.max(dataFiles[0].length(), dataFiles[1].length());
+        assertTrue("The file size should be close to 1MiB (with at most 50KiB error rate for the test)",
+                Math.abs(1024 * 1024 - closeTo1MiBFileSize) < 50 * 1024);
```
Comment (Maxwell-Guo) [inline test/unit/org/apache/cassandra/io/sstable/CQLSSTableWriterTest.java]: code format

**Comment by yifan-c** [comment]
Closing the committed patch

**Pull request closed without merging**